File: 算法设计与分析作业/3.14 完全平方数（BFS）.py
from queue import *
from math import *

class state:
    temp=0
    step=0
    vec=[]

n=int(input())
ans=0
list=[]
num=int(sqrt(n))
a=[i*i for i in range(1,num+1)]
q=Queue()
for i in range(num):
    t=state()
    t.temp=a[i]
    t.step=1
    t.vec=[]
    t.vec.append(a[i])
    q.put(t)
while not q.empty():
    s=q.get()
    if(s.temp==n):
        ans=s.step
        list=s.vec[:]
        break
    for i in range(num):
        if(s.temp+a[i]<=n):
            t=state()
            t.temp=s.temp+a[i]
            t.step=s.step+1
            t.vec=s.vec[:]
            t.vec.append(a[i])
            q.put(t)
print(ans)
len=len(list)
#print(n,"=",list[0])
print("{}={}".format(n,list[0]),end="")
for i in range(1,len):
    #print("+",list[i])
    print("+{}".format(list[i]),end="")

File: 算法设计与分析作业/第2章 递归与分治/3.14 完全平方数（BFS）.cpp
#include <iostream>
#include <cstdio>
#include <cmath>
#include <queue>
#include <vector>
using namespace std;
struct state
{
	int temp,step;//temp是当前的和，step就是当前搜索的步数，也就是当前需要的数字数量 
	//vector<int> v;
};
int main()
{
	int n,a[105],ans;
	//vector<int> out;
	cin>>n;
	int num=sqrt(n);
	for(int i=1;i<=num;i++)
		a[i]=i*i;
	queue<state> q;//BFS的队列 
	for(int i=1;i<=num;i++)
	{
		state t;
		t.temp=a[i];
		t.step=1;
		//t.v.push_back(a[i]);
		q.push(t);
	}
	while(!q.empty())
	{
		state s=q.front();
		q.pop();
		if(s.temp==n)
		{
			ans=s.step;
			//int size=s.v.size();
			//for(int i=0;i<size;i++)
			//	out.push_back(s.v[i]);
			break;
		}
		for(int i=1;i<=num;i++)
			if(s.temp+a[i]<=n)
			{
				state t;
				t.temp=s.temp+a[i];
				t.step=s.step+1;
				//t.v=s.v;
				//t.v.push_back(a[i]);
				q.push(t);
			}
	}
	printf("%d\n",ans);
	//printf("%d=%d",n,out[0]);
	//for(int i=1;i<out.size();i++)
	//	printf("+%d",out[i]);
	return 0;
}
